sdk/servicebus/azure-messaging-servicebus-track1-perf/src/main/java/com/microsoft/azure/servicebus/perf/ReceiveAndLockMessageTest.java

```diff
@@ -0,0 +1,95 @@
+
+package com.microsoft.azure.servicebus.perf;
+
+import com.azure.core.util.logging.ClientLogger;
+import com.microsoft.azure.servicebus.perf.core.ServiceBusStressOptions;
+import com.microsoft.azure.servicebus.perf.core.ServiceTest;
+import com.microsoft.azure.servicebus.IMessage;
+import com.microsoft.azure.servicebus.Message;
+import com.microsoft.azure.servicebus.ReceiveMode;
+import com.microsoft.azure.servicebus.primitives.ServiceBusException;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Performance test.
+ */
+public class ReceiveAndLockMessageTest extends ServiceTest<ServiceBusStressOptions> {
+    private final ClientLogger logger = new ClientLogger(ReceiveAndLockMessageTest.class);
+    private final ServiceBusStressOptions options;
+
+    /**
+     * Creates test object
+     * @param options to set performance test options.
+     */
+    public ReceiveAndLockMessageTest(ServiceBusStressOptions options) {
+        super(options, ReceiveMode.PEEKLOCK);
+        this.options = options;
+    }
+
+    private Mono<Void> sendMessage() {
+        int total = options.getParallel() * options.getMessagesToSend() * TOTAL_MESSAGE_MULTIPLIER;
+        List<Message> messages = new ArrayList<>();
+        for (int i = 0; i < total; ++i) {
+            Message message = new Message(CONTENTS);
+            message.setMessageId(UUID.randomUUID().toString());
+            messages.add(message);
+        }
+        return Mono.fromFuture(sender.sendBatchAsync(messages));
+    }
+
+    @Override
+    public Mono<Void> globalSetupAsync() {
+        // Since test does warm up and test many times, we are sending many messages, so we will have them available.
+        return super.globalSetupAsync()
+            .then(sendMessage());
+    }
+
+    @Override
+    public void run() {
+        Collection<IMessage> messages;
+        try {
+            messages = receiver.receiveBatch(options.getMessagesToReceive());
+        } catch (Exception e) {
+            throw logger.logExceptionAsWarning(new RuntimeException(e));
+        }
+
+        for (IMessage message : messages) {
+            try {
+                receiver.complete(message.getLockToken());
+            } catch (InterruptedException | ServiceBusException e) {
+                throw logger.logExceptionAsWarning(new RuntimeException(e));
+            }
+        }
+    }
+
+    @Override
+    public Mono<Void> runAsync() {
+        CompletableFuture<Void> completeTask = CompletableFuture.runAsync(() -> {
+        });
+        return Mono.fromFuture(receiver.receiveBatchAsync(options.getMessagesToReceive())
+            .thenComposeAsync(iMessages -> {
+                for (IMessage message : iMessages) {
+                    completeTask.thenComposeAsync(ignore -> receiver.completeAsync(message.getLockToken()));
```

Comment:
we shouldn't create our async threads here using CompletableFutures.
The whole block should execute synchronously in the logic within runAsync.

Reply:
The complete operation is asynchronous because it has to make a network call. The problem here is that we're not chaining downstream operators correctly.